src/eos-updater-apply.c

```diff
@@ -73,41 +77,506 @@ get_test_osname (void)
+       *
+       * We are not able to meaningfully clean up here - the refs will remain
+       * in the flatpak ostree repo for the next time we want to install them
+       * but there's nothing in the public API that will allow us to get rid of
+       * them. */
+      if (!g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_ONLY_PULLED))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          cleanup_undeployed_flatpaks_in_installation (installation, flatpaks_to_deploy, i);
+          return FALSE;
+        }
+    }
+
+  return TRUE;
+}
+
+static const gchar *
+get_pending_flatpak_deployments_state_path (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_UPDATER_FLATPAK_UPGRADE_STATE_DIR",
+                                    LOCALSTATEDIR "/lib/eos-application-tools");
+}
+
+/* We return a transfer-full GFile of the record of flatpaks to be
+ * deployed on the next boot. If something goes wrong the caller is expected
+ * to delete the file. In any event, the caller must free the return value. */
+static GFile *
+write_flatpaks_to_be_deployed (GPtrArray    *flatpaks_to_deploy,
+                               const gchar  *deployment_csum,
+                               GCancellable *cancellable,
+                               GError       **error)
+{
+  const gchar *pending_flatpak_deployments_state_path = get_pending_flatpak_deployments_state_path ();
+  g_autofree gchar *pending_flatpak_deployments_to_export_dir_path = g_build_filename (pending_flatpak_deployments_state_path,
+                                                                                       deployment_csum,
+                                                                                       NULL);
+  g_autofree gchar *pending_flatpak_deployments_path = g_build_filename (pending_flatpak_deployments_to_export_dir_path,
+                                                                         "to-export",
+                                                                         NULL);
+  g_autoptr(GFile) pending_flatpak_deployments_to_export_dir = g_file_new_for_path (pending_flatpak_deployments_to_export_dir_path);
+  g_auto(GStrv) flatpak_names_to_autodeploy = g_new0 (gchar *, flatpaks_to_deploy->len + 1);
+  g_autofree gchar *flatpak_names_to_autodeploy_contents = NULL;
+  g_autoptr(GError) local_error = NULL;
+  gsize i = 0;
+
+  for (; i < flatpaks_to_deploy->len; ++i)
+    {
+      FlatpakRemoteRef *to_install = FLATPAK_REMOTE_REF (g_ptr_array_index (flatpaks_to_deploy, i));
+      g_autofree gchar *flatpak_ref = flatpak_ref_format_ref (FLATPAK_REF (to_install));
+      flatpak_names_to_autodeploy[i] = g_strdup_printf ("%s:%s",
+                                                        flatpak_remote_ref_get_remote_name (to_install),
+                                                        flatpak_ref);
+    }
+
+  flatpak_names_to_autodeploy_contents = g_strjoinv ("\n", flatpak_names_to_autodeploy);
```

Comment:
As another optional suggestion, you could also use e.g. `g_file_replace()` and https://developer.gnome.org/gio/stable/GDataOutputStream.html#g-data-output-stream-put-string to build the new file instead of writing it all at once and manage temporary buffers.

Reply:
It is probably better to either have the entire file or not have it at all in this case. Though I am not sure what the contents of the underlying file are defined to be if `g_file_set_contents` fails